agent/engine/docker_task_engine.go

```diff
@@ -1534,25 +1535,91 @@ func (engine *DockerTaskEngine) pullAndUpdateContainerReference(task *apitask.Ta
 		defer clearCreds()
 	}
 
-	metadata := engine.client.PullImage(engine.ctx, container.Image, container.RegistryAuthentication, engine.cfg.ImagePullTimeout)
+	imageRef := container.Image
+	imageDigest := container.GetImageDigest()
+	if imageDigest != "" {
+		// If image digest is available then prepare a canonical reference to be used for image pull.
+		// This ensures that the image version referenced by the digest is pulled.
+		canonicalRef, err := referenceutil.GetCanonicalRef(imageRef, imageDigest)
+		if err != nil {
+			logger.Error("Failed to prepare a canonical reference. Cannot pull image.", logger.Fields{
+				field.TaskID:      task.GetID(),
+				field.Container:   container.Name,
+				field.Image:       imageRef,
+				field.ImageDigest: imageDigest,
+				field.Error:       err,
+			})
+			return dockerapi.DockerContainerMetadata{
+				Error: dockerapi.CannotPullContainerError{
+					FromError: fmt.Errorf(
+						"failed to prepare a canonical reference with image '%s' and digest '%s': %w",
+						imageRef, imageDigest, err),
+				},
+			}
+		}
+		imageRef = canonicalRef.String()
+		logger.Info("Prepared a canonical reference for image pull", logger.Fields{
+			field.TaskID:      task.GetID(),
+			field.Container:   container.Name,
+			field.Image:       container.Image,
+			field.ImageDigest: imageDigest,
+			field.ImageRef:    imageRef,
+		})
+	}
+
+	metadata := engine.client.PullImage(engine.ctx, imageRef, container.RegistryAuthentication, engine.cfg.ImagePullTimeout)
 
 	// Don't add internal images(created by ecs-agent) into imagemanger state
 	if container.IsInternal() {
 		return metadata
 	}
 	pullSucceeded := metadata.Error == nil
+
+	if pullSucceeded && imageRef != container.Image {
+		// Resolved image manifest digest was used to pull the image.
+		// Tag the pulled image so that it can be found using the image reference in the task.
+		ctx, cancel := context.WithTimeout(engine.ctx, tagImageTimeout)
+		defer cancel()
+		err := engine.client.TagImage(ctx, imageRef, container.Image)
```

Comment:
Q: Will there ever be a case where the image is already tagged here? I'm assume no but was curious if there's a scenario where we need to call `renameImage()`.

Reply:
Disregard! This is only used in our integ tests.